### PR TITLE
feat(ts): Improve types for directives

### PR DESCRIPTION
**Pull request by dac09**
### What does this do? 

- Improves (and corrects) the types for `ValidatorDirectiveFunc` and `TransformerDirectiveFunc` . 
- Updates templates for transformer and validator directives
- Updates the requireAuth directive in template
- Adds docs to typescript/utility-types about this too

### CODE REVIEW

File: docs/docs/typescript/utility-types.md [hunk 1]
```diff
@@ -220,3 +220,68 @@ export const handler = async (
...
+
+```ts
+// 👇 makes sure the functions' arguments are typed
+// highlight-next-line
+const transform: TransformerDirectiveFunc = ({ context, resolvedValue }) => {
+```
+
+This type takes a two generics - the type of the field you are transforming, and the type of your `directiveArgs`.
+
+So for example, let's say you have a transformer directive `@maskedEmail(permittedRoles: ['ADMIN']` that you apply to `String` fields. You would pass in the following types
```
Comment (Tobbe) [inline docs/docs/typescript/utility-types.md:276]:
```suggestion
So for example, let's say you have a transformer directive `@maskedEmail(permittedRoles: ['ADMIN'])` that you apply to `String` fields. You would pass in the following types
```

File: packages/graphql-server/src/plugins/useRedwoodDirective.ts [hunk 2]
```diff
@@ -13,39 +13,73 @@ import {
...
+ * @example ValidatorDirectiveFunc<{ roles?: string[] }>
+ *
  */
-export type ValidatorDirectiveFunc<FieldType = any> = (
-  args: Omit<DirectiveParams<FieldType>, 'resolvedValue'>
+export type ValidatorDirectiveFunc<TDirectiveArgs = Record<string, any>> = (
+  args: ValidateArgs<TDirectiveArgs>
 ) => Promise<void> | void
 
+
```
Comment (Tobbe) [inline packages/graphql-server/src/plugins/useRedwoodDirective.ts:56]:
```suggestion
```

**Comment by dac09** [comment]
@Tobbe  go / no go?

File: docs/docs/typescript/utility-types.md [hunk 3]
```diff
@@ -220,3 +220,68 @@ export const handler = async (
...
+### `TransformerDirectiveFunc`
+When you generate a [transformer directive](directives.md#transformers) you will see your `transform` function typed with `TransformDirectiveFunc<TField, TDirectiveArgs>`.
+
+```ts
+// 👇 makes sure the functions' arguments are typed
+// highlight-next-line
+const transform: TransformerDirectiveFunc = ({ context, resolvedValue }) => {
+```
+
+This type takes a two generics - the type of the field you are transforming, and the type of your `directiveArgs`.
```
Comment (Tobbe) [inline docs/docs/typescript/utility-types.md:274]:
```suggestion
This type takes two generics - the type of the field you are transforming, and the type of your `directiveArgs`.
```

**Pull request merged**